logutils/logutils.go

```diff
@@ -255,42 +306,243 @@ func shouldSkipFrame(frame runtime.Frame) bool {
+				currentNumDroppedLogs-numSeenDroppedLogs))
+		}
+		numSeenDroppedLogs = currentNumDroppedLogs
+
+		err := d.write(ql)
+		if err != nil {
+			counterLogErrors.Inc()
+			fmt.Fprintf(os.Stderr, "Failed to write to syslog: %v", err)
+		}
+		ql.OnLogDone()
+	}
+}
+
+func (d *SyslogDestination) write(ql QueuedLog) error {
+	switch ql.Level {
+	case log.PanicLevel:
+		return d.writer.Crit(ql.SyslogMessage)
+	case log.FatalLevel:
+		return d.writer.Crit(ql.SyslogMessage)
+	case log.ErrorLevel:
+		return d.writer.Err(ql.SyslogMessage)
+	case log.WarnLevel:
+		return d.writer.Warning(ql.SyslogMessage)
+	case log.InfoLevel:
+		return d.writer.Info(ql.SyslogMessage)
+	case log.DebugLevel:
+		return d.writer.Debug(ql.SyslogMessage)
+	default:
+		return nil
+	}
+}
+
+type Destination interface {
+	Level() log.Level
+	LoopWritingLogs()
+	Channel() chan<- QueuedLog
+	OnLogDropped()
+}
+
+type BackgroundHook struct {
+	levels      []log.Level
+	syslogLevel log.Level
+
+	destinations []Destination
+
+	// Our own copy of the dropped logs counter, used for logging out when we drop logs.
+	// Must be read/updated using atomic.XXX.
+	numDroppedLogs  uint64
+	lastDropLogTime time.Duration
+}
+
+func NewBackgroundHook(levels []log.Level, syslogLevel log.Level, destinations []Destination) *BackgroundHook {
+	return &BackgroundHook{
+		destinations: destinations,
+		levels:       levels,
+		syslogLevel:  syslogLevel,
+	}
+}
+
+func (h *BackgroundHook) Levels() []log.Level {
 	return h.levels
 }
 
-func (h *StreamHook) Fire(entry *log.Entry) (err error) {
+func (h *BackgroundHook) Fire(entry *log.Entry) (err error) {
 	var serialized []byte
 	if serialized, err = entry.Logger.Formatter.Format(entry); err != nil {
 		return
 	}
 
-	h.mu.Lock()
-	defer h.mu.Unlock()
+	// entry's buffer will be reused after we return but we're about to send the message over
+	// a channel so we need to take a copy.
+	bufCopy := make([]byte, len(serialized))
+	copy(bufCopy, serialized)
+	if entry.Buffer != nil {
+		entry.Buffer.Truncate(0)
+	}
 
-	_, err = h.writer.Write(serialized)
-	return
-}
+	ql := QueuedLog{
+		Level:   entry.Level,
+		Message: bufCopy,
+	}
 
-type LeveledHook struct {
-	hook   log.Hook
-	levels []log.Level
-}
+	if entry.Level <= h.syslogLevel {
+		// syslog gets its own log string since our default log string duplicates a lot of
+		// syslog metadata.  Only calculate that string if it's needed.
+		ql.SyslogMessage = FormatForSyslog(entry)
+	}
 
-func (h *LeveledHook) Levels() []log.Level {
-	return h.levels
+	var waitGroup *sync.WaitGroup
+	if entry.Level <= log.FatalLevel {
+		// This is a panic or a fatal log so we'll get killed immediately after we return.
+		// Use a wait group to wait for the log to get written.
+		waitGroup = &sync.WaitGroup{}
+		ql.WaitGroup = waitGroup
+	}
+
+	for _, dest := range h.destinations {
+		if ql.Level > dest.Level() {
+			continue
+		}
+		if waitGroup != nil {
+			waitGroup.Add(1)
+		}
+		select {
+		case dest.Channel() <- ql:
```

Comment:
Sadly, that's not safe, you have to increment it before sending it to the background thread (or the background thread could get scheduled immediately and decrement it before foreground increments, which fails).